#!/bin/bash
 
 #SBATCH -t 0-24:00 
 #SBATCH -c 20 #number of cpus required per task
 #SBATCH -n 10 #launch a maximum of number tasks 
 #SBATCH --mem 60G
 #SBATCH --job-name STAR_align_1
 #SBATCH -o %j.out  #file to which standard out will be written
 #SBATCH -e %j.err
 
 module load STAR/2.6.1c
 
/data/programs/STAR/2.6.1c_2018-10-30/bin/Linux_x86_64/STAR --runThreadN 20 --readFilesIn /liulab/jjyu/DB_bulkRNAseq/BDU1_1.fq.gz /liulab/jjyu/DB_bulkRNAseq/BDU1_2.fq.gz --quantMode TranscriptomeSAM --outSAMtype BAM SortedByCoordinate --outFileNamePrefix /liulab/jjyu/DB_bulkRNAseq/STAR_result/BDU1 --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --alignMatesGapMax 1000000 --chimSegmentMin 20 --readFilesCommand zcat --genomeDir /liulab/jjyu/Reference/hg38_index &
 
/data/programs/STAR/2.6.1c_2018-10-30/bin/Linux_x86_64/STAR --runThreadN 20 --readFilesIn /liulab/jjyu/DB_bulkRNAseq/BDU2_1.fq.gz /liulab/jjyu/DB_bulkRNAseq/BDU2_2.fq.gz --quantMode TranscriptomeSAM --outSAMtype BAM SortedByCoordinate --outFileNamePrefix /liulab/jjyu/DB_bulkRNAseq/STAR_result/BDU2 --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --alignMatesGapMax 1000000 --chimSegmentMin 20 --readFilesCommand zcat --genomeDir /liulab/jjyu/Reference/hg38_index &
  
/data/programs/STAR/2.6.1c_2018-10-30/bin/Linux_x86_64/STAR --runThreadN 20 --readFilesIn /liulab/jjyu/DB_bulkRNAseq/BDU3_1.fq.gz /liulab/jjyu/DB_bulkRNAseq/BDU3_2.fq.gz --quantMode TranscriptomeSAM --outSAMtype BAM SortedByCoordinate --outFileNamePrefix /liulab/jjyu/DB_bulkRNAseq/STAR_result/BDU3 --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --alignMatesGapMax 1000000 --chimSegmentMin 20 --readFilesCommand zcat --genomeDir /liulab/jjyu/Reference/hg38_index &
    
/data/programs/STAR/2.6.1c_2018-10-30/bin/Linux_x86_64/STAR --runThreadN 20 --readFilesIn /liulab/jjyu/DB_bulkRNAseq/BDV1_1.fq.gz /liulab/jjyu/DB_bulkRNAseq/BDV1_2.fq.gz --quantMode TranscriptomeSAM --outSAMtype BAM SortedByCoordinate --outFileNamePrefix /liulab/jjyu/DB_bulkRNAseq/STAR_result/BDV1 --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --alignMatesGapMax 1000000 --chimSegmentMin 20 --readFilesCommand zcat --genomeDir /liulab/jjyu/Reference/hg38_index &

/data/programs/STAR/2.6.1c_2018-10-30/bin/Linux_x86_64/STAR --runThreadN 20 --readFilesIn /liulab/jjyu/DB_bulkRNAseq/BDV2_1.fq.gz /liulab/jjyu/DB_bulkRNAseq/BDV2_2.fq.gz --quantMode TranscriptomeSAM --outSAMtype BAM SortedByCoordinate --outFileNamePrefix /liulab/jjyu/DB_bulkRNAseq/STAR_result/BDV2 --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --alignMatesGapMax 1000000 --chimSegmentMin 20 --readFilesCommand zcat --genomeDir /liulab/jjyu/Reference/hg38_index &

/data/programs/STAR/2.6.1c_2018-10-30/bin/Linux_x86_64/STAR --runThreadN 20 --readFilesIn /liulab/jjyu/DB_bulkRNAseq/BDV3_1.fq.gz /liulab/jjyu/DB_bulkRNAseq/BDV3_2.fq.gz --quantMode TranscriptomeSAM --outSAMtype BAM SortedByCoordinate --outFileNamePrefix /liulab/jjyu/DB_bulkRNAseq/STAR_result/BDV3 --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --alignMatesGapMax 1000000 --chimSegmentMin 20 --readFilesCommand zcat --genomeDir /liulab/jjyu/Reference/hg38_index &

/data/programs/STAR/2.6.1c_2018-10-30/bin/Linux_x86_64/STAR --runThreadN 20 --readFilesIn /liulab/jjyu/DB_bulkRNAseq/BDV4_1.fq.gz /liulab/jjyu/DB_bulkRNAseq/BDV4_2.fq.gz --quantMode TranscriptomeSAM --outSAMtype BAM SortedByCoordinate --outFileNamePrefix /liulab/jjyu/DB_bulkRNAseq/STAR_result/BDV4 --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --alignMatesGapMax 1000000 --chimSegmentMin 20 --readFilesCommand zcat --genomeDir /liulab/jjyu/Reference/hg38_index &

/data/programs/STAR/2.6.1c_2018-10-30/bin/Linux_x86_64/STAR --runThreadN 20 --readFilesIn /liulab/jjyu/DB_bulkRNAseq/BDV5_1.fq.gz /liulab/jjyu/DB_bulkRNAseq/BDV5_2.fq.gz --quantMode TranscriptomeSAM --outSAMtype BAM SortedByCoordinate --outFileNamePrefix /liulab/jjyu/DB_bulkRNAseq/STAR_result/BDV5 --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --alignMatesGapMax 1000000 --chimSegmentMin 20 --readFilesCommand zcat --genomeDir /liulab/jjyu/Reference/hg38_index &

/data/programs/STAR/2.6.1c_2018-10-30/bin/Linux_x86_64/STAR --runThreadN 20 --readFilesIn /liulab/jjyu/DB_bulkRNAseq/BDV6_1.fq.gz /liulab/jjyu/DB_bulkRNAseq/BDV6_2.fq.gz --quantMode TranscriptomeSAM --outSAMtype BAM SortedByCoordinate --outFileNamePrefix /liulab/jjyu/DB_bulkRNAseq/STAR_result/BDV6 --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --alignMatesGapMax 1000000 --chimSegmentMin 20 --readFilesCommand zcat --genomeDir /liulab/jjyu/Reference/hg38_index 