#!/bin/bash
 
 #SBATCH -t 0-24:00 
 #SBATCH -c 20 #number of cpus required per task
 #SBATCH -n 10 #launch a maximum of number tasks 
 #SBATCH --mem 100G
 #SBATCH --job-name STAR_align_1
 #SBATCH -o %j.out  #file to which standard out will be written
 #SBATCH -e %j.err
 
 module load STAR/2.6.1c
 
/data/programs/STAR/2.6.1c_2018-10-30/bin/Linux_x86_64/STAR --runThreadN 20 --readFilesIn /liulab/jjyu/DB_bulkRNAseq/HC1_1.fq.gz /liulab/jjyu/DB_bulkRNAseq/HC1_2.fq.gz --quantMode TranscriptomeSAM --outSAMtype BAM SortedByCoordinate --outFileNamePrefix /liulab/jjyu/DB_bulkRNAseq/STAR_result/HC1 --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --alignMatesGapMax 1000000 --chimSegmentMin 20 --readFilesCommand zcat --genomeDir /liulab/jjyu/Reference/hg38_index 
 
 /data/programs/STAR/2.6.1c_2018-10-30/bin/Linux_x86_64/STAR --runThreadN 20 --readFilesIn /liulab/jjyu/DB_bulkRNAseq/HC2_1.fq.gz /liulab/jjyu/DB_bulkRNAseq/HC2_2.fq.gz --quantMode TranscriptomeSAM --outSAMtype BAM SortedByCoordinate --outFileNamePrefix /liulab/jjyu/DB_bulkRNAseq/STAR_result/HC2 --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --alignMatesGapMax 1000000 --chimSegmentMin 20 --readFilesCommand zcat --genomeDir /liulab/jjyu/Reference/hg38_index &
 
 /data/programs/STAR/2.6.1c_2018-10-30/bin/Linux_x86_64/STAR --runThreadN 20 --readFilesIn /liulab/jjyu/DB_bulkRNAseq/HC3_1.fq.gz /liulab/jjyu/DB_bulkRNAseq/HC3_2.fq.gz --quantMode TranscriptomeSAM --outSAMtype BAM SortedByCoordinate --outFileNamePrefix /liulab/jjyu/DB_bulkRNAseq/STAR_result/HC3 --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --alignMatesGapMax 1000000 --chimSegmentMin 20 --readFilesCommand zcat --genomeDir /liulab/jjyu/Reference/hg38_index &

 /data/programs/STAR/2.6.1c_2018-10-30/bin/Linux_x86_64/STAR --runThreadN 20 --readFilesIn /liulab/jjyu/DB_bulkRNAseq/HC4_1.fq.gz /liulab/jjyu/DB_bulkRNAseq/HC4_2.fq.gz --quantMode TranscriptomeSAM --outSAMtype BAM SortedByCoordinate --outFileNamePrefix /liulab/jjyu/DB_bulkRNAseq/STAR_result/HC4 --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --alignMatesGapMax 1000000 --chimSegmentMin 20 --readFilesCommand zcat --genomeDir /liulab/jjyu/Reference/hg38_index &
 
  /data/programs/STAR/2.6.1c_2018-10-30/bin/Linux_x86_64/STAR --runThreadN 20 --readFilesIn /liulab/jjyu/DB_bulkRNAseq/HC5_1.fq.gz /liulab/jjyu/DB_bulkRNAseq/HC5_2.fq.gz --quantMode TranscriptomeSAM --outSAMtype BAM SortedByCoordinate --outFileNamePrefix /liulab/jjyu/DB_bulkRNAseq/STAR_result/HC5 --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --alignMatesGapMax 1000000 --chimSegmentMin 20 --readFilesCommand zcat --genomeDir /liulab/jjyu/Reference/hg38_index &
  
  /data/programs/STAR/2.6.1c_2018-10-30/bin/Linux_x86_64/STAR --runThreadN 20 --readFilesIn /liulab/jjyu/DB_bulkRNAseq/HC6_1.fq.gz /liulab/jjyu/DB_bulkRNAseq/HC6_2.fq.gz --quantMode TranscriptomeSAM --outSAMtype BAM SortedByCoordinate --outFileNamePrefix /liulab/jjyu/DB_bulkRNAseq/STAR_result/HC6 --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --alignMatesGapMax 1000000 --chimSegmentMin 20 --readFilesCommand zcat --genomeDir /liulab/jjyu/Reference/hg38_index &
 
 /data/programs/STAR/2.6.1c_2018-10-30/bin/Linux_x86_64/STAR --runThreadN 20 --readFilesIn /liulab/jjyu/DB_bulkRNAseq/HC7_1.fq.gz /liulab/jjyu/DB_bulkRNAseq/HC7_2.fq.gz --quantMode TranscriptomeSAM --outSAMtype BAM SortedByCoordinate --outFileNamePrefix /liulab/jjyu/DB_bulkRNAseq/STAR_result/HC7 --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --alignMatesGapMax 1000000 --chimSegmentMin 20 --readFilesCommand zcat --genomeDir /liulab/jjyu/Reference/hg38_index &
 
 /data/programs/STAR/2.6.1c_2018-10-30/bin/Linux_x86_64/STAR --runThreadN 20 --readFilesIn /liulab/jjyu/DB_bulkRNAseq/HC8_1.fq.gz /liulab/jjyu/DB_bulkRNAseq/HC8_2.fq.gz --quantMode TranscriptomeSAM --outSAMtype BAM SortedByCoordinate --outFileNamePrefix /liulab/jjyu/DB_bulkRNAseq/STAR_result/HC8 --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --alignMatesGapMax 1000000 --chimSegmentMin 20 --readFilesCommand zcat --genomeDir /liulab/jjyu/Reference/hg38_index &

 /data/programs/STAR/2.6.1c_2018-10-30/bin/Linux_x86_64/STAR --runThreadN 20 --readFilesIn /liulab/jjyu/DB_bulkRNAseq/HC9_1.fq.gz /liulab/jjyu/DB_bulkRNAseq/HC9_2.fq.gz --quantMode TranscriptomeSAM --outSAMtype BAM SortedByCoordinate --outFileNamePrefix /liulab/jjyu/DB_bulkRNAseq/STAR_result/HC9 --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --alignMatesGapMax 1000000 --chimSegmentMin 20 --readFilesCommand zcat --genomeDir /liulab/jjyu/Reference/hg38_index &
 
  /data/programs/STAR/2.6.1c_2018-10-30/bin/Linux_x86_64/STAR --runThreadN 20 --readFilesIn /liulab/jjyu/DB_bulkRNAseq/HC10_1.fq.gz /liulab/jjyu/DB_bulkRNAseq/HC10_2.fq.gz --quantMode TranscriptomeSAM --outSAMtype BAM SortedByCoordinate --outFileNamePrefix /liulab/jjyu/DB_bulkRNAseq/STAR_result/HC10 --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --alignMatesGapMax 1000000 --chimSegmentMin 20 --readFilesCommand zcat --genomeDir /liulab/jjyu/Reference/hg38_index 